pkg/schema/definitions/handler.go

```diff
@@ -130,13 +147,52 @@ func (s *SchemaDefinitionHandler) indexSchemaNames(models proto.Models, groups *
 			// we can safely continue
 			continue
 		}
+		schemaID := converter.GVKToSchemaID(*gvk)
 		prefVersion := preferredResourceVersions[gvk.Group]
-		// if we don't have a known preferred version for this group or we are the preferred version
-		// add this as the model name for the schema
+		// we always add the preferred version to the map. However, if this isn't the preferred version the preferred group could
+		// be missing this resource (e.x. v1alpha1 has a resource, it's removed in v1). In those cases, we add the model name
+		// only if we don't already have an entry. This way we always choose the preferred, if possible, but still have 1 version
+		// for everything
 		if prefVersion == "" || prefVersion == gvk.Version {
-			schemaID := converter.GVKToSchemaID(*gvk)
 			schemaToModel[schemaID] = modelName
+		} else {
+			_, ok := schemaToModel[schemaID]
+			if !ok {
+				schemaToModel[schemaID] = modelName
+			}
 		}
 	}
 	return schemaToModel
 }
+
+// baseSchemaToDefinition converts a given schema to the definition map. This should only be used with baseSchemas, whose definitions
+// are expected to be set by another application and may not be k8s resources.
+func baseSchemaToDefinition(schema types.APISchema) map[string]definition {
+	definitions := map[string]definition{}
+	def := definition{
+		Description:    schema.Description,
+		Type:           schema.ID,
+		ResourceFields: map[string]definitionField{},
+	}
+	for fieldName, field := range schema.ResourceFields {
+		fieldType, subType := parseFieldType(field.Type)
+		def.ResourceFields[fieldName] = definitionField{
+			Type:        fieldType,
+			SubType:     subType,
+			Description: field.Description,
+			Required:    field.Required,
+		}
+	}
+	definitions[schema.ID] = def
+	return definitions
+}
+
+// parseFieldType parses a schemas.Field's type to a type (first return) and subType (second return)
+func parseFieldType(fieldType string) (string, string) {
+	splitted := strings.Split(fieldType, "[")
+	if len(splitted) == 2 && len(splitted[1]) > 0 {
+		// example: map[someVal] - split removes the leading "[" trim removes the trailing "]"
+		return splitted[0], strings.TrimSuffix(splitted[1], "]")
+	}
+	return fieldType, ""
+}
```

Comment:
This somewhat duplicates the functionality here: https://github.com/rancher/wrangler/blob/v2.2.0-rc6/pkg/schemas/definition/definition.go#L25. Wrangler doesn't have a function to get both types and subtypes though, a bit unfortunate.

Though a mix of `wrangler.Subtype` and `wrangler.Is{Map,Array,Reference}Type` should do the trick.

WDYT?

Reply:
I'm a bit split on it. I don't want to re-implement logic, but I also don't want to make a longer, worse-performing function in order to avoid re-implementing ~5 lines. Let me try implementing a replacement function for this and see how that looks.

Reply:
I've update the function to use this approach. It might have slightly less performance but I think it reads a bit cleaner and avoid us doing a string split here. LMK what you think.